pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -288,33 +297,65 @@ func (r *SubnetSetReconciler) CollectGarbage(ctx context.Context) {
 	}
 }
 
-func (r *SubnetSetReconciler) DeleteSubnetForSubnetSet(obj v1alpha1.SubnetSet, updataStatus bool) (bool, error) {
-	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetSetCRUID, string(obj.GetUID()))
-	hitError := false
-	hasStaleSubnetPorts := false
+func (r *SubnetSetReconciler) cleanStaleSubnetsForSubnetSet(subnetSetName, ns, subnetsetID string) error {
+	nsxStaleSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetSetCRNamespacedName, util.CombineNamespaceName(subnetSetName, ns))
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetSetCRUID, subnetsetID)
+	subnetIDs := sets.New[string]()
 	for _, subnet := range nsxSubnets {
-		r.SubnetService.LockSubnet(subnet.Path)
-		portNums := len(r.SubnetPortService.GetPortsOfSubnet(*subnet.Id))
-		if portNums > 0 {
-			hasStaleSubnetPorts = true
-			r.SubnetService.UnlockSubnet(subnet.Path)
-			continue
-		}
-		if err := r.SubnetService.DeleteSubnet(*subnet); err != nil {
-			log.Error(err, "fail to delete subnet from subnetset cr", "ID", *subnet.Id)
-			hitError = true
+		subnetIDs.Insert(*subnet.Id)
+	}
+	subnetToDelete := []*model.VpcSubnet{}
+	for _, staleSubnet := range nsxStaleSubnets {
+		if !subnetIDs.Has(*staleSubnet.Id) {
+			subnetToDelete = append(subnetToDelete, staleSubnet)
 		}
-		r.SubnetService.UnlockSubnet(subnet.Path)
 	}
-	if updataStatus {
-		if err := r.SubnetService.UpdateSubnetSetStatus(&obj); err != nil {
-			return hasStaleSubnetPorts, err
+	if err := r.deleteSubnets(subnetToDelete); err != nil {
+		log.Error(err, "Failed to delete stale subnets", "SubnetSet", subnetSetName, "Namespace", ns)
+		return err
+	}
+	log.Info("Successfully cleaned stale subnets for SubnetSet", "SubnetSet", subnetSetName, "Namespace", ns, "num", len(subnetToDelete))
+	return nil
+}
+
+func (r *SubnetSetReconciler) deleteSubnetBySubnetSetName(subnetSetName, ns string) error {
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetSetCRNamespacedName, util.CombineNamespaceName(subnetSetName, ns))
+	if err := r.deleteSubnets(nsxSubnets); err != nil {
```

Comment:
Shall we also check if those nsx subnets belongs to an existing subnetset with the same name/namespace by fetching all the existing subnetset cr uid, and the corresponding subnets uid to filter these subnets, which is similar to what we do in subnet controller?

Reply:
Right, that is true. It's better to move `deleteSubnets` to service for sharing function between subnet and subnetset controller. 
We also need get some K8s CR ids, these info can be passed from controller to service when calling `deleteSubnets` 

but this is a sort of refactoring work, we can refine it in the next patch.